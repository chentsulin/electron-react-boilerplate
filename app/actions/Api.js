// @flow
import type { counterStateType } from '../reducers/Api';

type actionType = {
  +type: string
};

export const INCREMENT_COUNTER = 'INCREMENT_COUNTER';
export const DECREMENT_COUNTER = 'DECREMENT_COUNTER';

export function increment() {
  return {
    type: INCREMENT_COUNTER
  };
}

export function decrement() {
  return {
    type: DECREMENT_COUNTER
  };
}

export function incrementIfOdd() {
  return (dispatch: (action: actionType) => void, getState: () => counterStateType) => {
    const { counter } = getState();

    if (counter % 2 === 0) {
      return;
    }

    dispatch(increment());
  };
}

export function incrementAsync(delay: number = 1000) {
  return (dispatch: (action: actionType) => void) => {
    setTimeout(() => {
      dispatch(increment());
    }, delay);
  };
}

// nicks new stuff
export function logResult(result) {
  console.log(result);
}

export function logError(error) {
  console.log('Looks like there was a problem: \n', error);
}

export function validateResponse(response) {
  if (!response.ok) {
    throw Error(response.statusText);
  }
  return response;
}

export function readResponseAsJSON(response) {
  return response.json();
}

export function fetchJSON(pathToResource) {
  fetch(pathToResource) // 1
  .then(validateResponse) // 2
  .then(readResponseAsJSON) // 3
  .then(logResult) // 4
  .catch(logError);
}
