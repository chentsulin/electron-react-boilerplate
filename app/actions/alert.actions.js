import { alertConstants } from '../constants';

export const alertActions = {
  success,
  error,
  clear
};

export default alertActions;

function success(message) {
  return { type: alertConstants.SUCCESS, message };
}

function error(message) {
  return { type: alertConstants.ERROR, message };
}

function clear() {
  return { type: alertConstants.CLEAR };
}
