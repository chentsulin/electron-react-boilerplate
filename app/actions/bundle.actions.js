import sort from 'fast-sort';
import traverse from 'traverse';
import { bundleConstants } from '../constants/bundle.constants';
import { bundleService } from '../services/bundle.service';

export const bundleActions = {
  mockFetchAll,
  fetchAll,
  delete: removeBundle,
  downloadResources,
  requestSaveBundleTo,
  toggleModePauseResume,
  toggleSelectBundle
};

export default bundleActions;

export function mockFetchAll() {
  return dispatch => {
    dispatch(request());
    return new Promise(resolve => {
      const mockBundles = getMockBundles();
      resolve(mockBundles);
    }).then(bundles => dispatch(success(bundles)), error => dispatch(failure(error)));
  };

  function request() {
    return { type: bundleConstants.FETCH_REQUEST };
  }
  function success(bundles) {
    return { type: bundleConstants.FETCH_SUCCESS, bundles };
  }
  function failure(error) {
    return { type: bundleConstants.FETCH_FAILURE, error };
  }
}

export function fetchAll() {
  return dispatch => {
    dispatch(request());

    return bundleService
      .fetchAll()
      .then(bundles => dispatch(success(bundles)), error => dispatch(failure(error)));
  };

  function request() {
    return { type: bundleConstants.FETCH_REQUEST };
  }
  function success(bundles) {
    return { type: bundleConstants.FETCH_SUCCESS, bundles };
  }
  function failure(error) {
    return { type: bundleConstants.FETCH_FAILURE, error };
  }
}

export function downloadResources(id) {
  return async dispatch => {
    try {
      const manifestResourcePaths = await bundleService.getManifestResourcePaths(id);
      manifestResourcePaths.unshift('metadata.xml');
      await bundleService.downloadResources(id);
      dispatch(request(id, manifestResourcePaths));
    } catch (error) {
      dispatch(failure(id, error));
    }
  };
  function request(_id, manifestResourcePaths) {
    return { type: bundleConstants.DOWNLOAD_RESOURCES_REQUEST, id: _id, manifestResourcePaths };
  }
  function failure(_id, error) {
    return { type: bundleConstants.DOWNLOAD_RESOURCES_FAILURE, id, error };
  }
}


function removeBundle(id) {
  return dispatch => {
    dispatch(request(id));

    bundleService
      .delete(id)
      .then(() => {
        dispatch(success(id));
        return true;
      })
      .catch(error => {
        dispatch(failure(id, error));
        return true;
      });
  };

  function request(_id) {
    return { type: bundleConstants.DELETE_REQUEST, id: _id };
  }
  function success(_id) {
    return { type: bundleConstants.DELETE_SUCCESS, id: _id };
  }
  function failure(_id, error) {
    return { type: bundleConstants.DELETE_FAILURE, id: _id, error };
  }
}

export function requestSaveBundleTo(id, selectedFolder) {
  return async dispatch => {
    const bundleInfo = await bundleService.fetchById(id);
    const bundleBytesToSave = traverse(bundleInfo.store.file_info).reduce(addByteSize, 0);
    const resourcePaths = await bundleService.getResourcePaths(id);
    resourcePaths.unshift('metadata.xml');
    const resourcePathsProgress = resourcePaths.reduce((acc, resourcePath) => {
      acc[resourcePath] = 0;
      return acc;
    }, {});
    let bundleBytesSaved = 0;
    dispatch(request(id, selectedFolder, bundleBytesToSave, resourcePaths));
    resourcePaths.forEach(async resourcePath => {
      try {
        const downloadItem = await bundleService.requestSaveResourceTo(
          selectedFolder,
          id,
          resourcePath,
          (resourceTotalBytesSaved, resourceProgress) => {
            const originalResourceBytesTransferred = resourcePathsProgress[resourcePath];
            resourcePathsProgress[resourcePath] = resourceTotalBytesSaved;
            const bytesDiff = resourceTotalBytesSaved - originalResourceBytesTransferred;
            bundleBytesSaved += bytesDiff;
            if (resourceProgress && resourceProgress % 100 === 0) {
              dispatch(updated(
                id,
                resourcePath,
                resourceTotalBytesSaved,
                bundleBytesSaved,
                bundleBytesToSave
              ));
            }
          }
        );
        return downloadItem;
      } catch (error) {
        dispatch(failure(id, error));
      }
    });
  };

  function addByteSize(accBytes, fileInfoNode) {
    if (fileInfoNode.is_dir || this.isRoot || fileInfoNode.size === undefined) {
      return accBytes;
    }
    return accBytes + fileInfoNode.size;
  }

  function request(_id, _folderName, bundleBytesToSave, resourcePaths) {
    return {
      type: bundleConstants.SAVETO_REQUEST,
      id: _id,
      folderName: _folderName,
      bundleBytesToSave,
      resourcePaths
    };
  }

  function updated(
    _id,
    resourcePath,
    resourceTotalBytesSaved,
    bundleBytesSaved,
    bundleBytesToSave
  ) {
    return {
      type: bundleConstants.SAVETO_UPDATED,
      id,
      resourcePath,
      resourceTotalBytesSaved,
      bundleBytesSaved,
      bundleBytesToSave
    };
  }
  function failure(_id, error) {
    return { type: bundleConstants.SAVETO_FAILURE, id: _id, error };
  }
}

export function toggleModePauseResume(id) {
  return { type: bundleConstants.TOGGLE_MODE_PAUSE_RESUME, id };
}

export function toggleSelectBundle(id) {
  return { type: bundleConstants.TOGGLE_SELECT, id };
}

function getMockBundles() {
  const bundles = [
    {
      id: 'bundle01',
      name: 'Test Bundle #1',
      revision: 3,
      task: 'UPLOAD',
      status: 'COMPLETED'
    },
    {
      id: 'bundle02',
      name: 'Another Bundle',
      revision: 3,
      task: 'UPLOAD',
      status: 'IN_PROGRESS',
      progress: 63,
      mode: 'PAUSED'
    },
    {
      id: 'bundle03',
      name: 'Audio Bundle',
      revision: 52,
      task: 'DOWNLOAD',
      status: 'IN_PROGRESS',
      progress: 12,
      mode: 'RUNNING'
    },
    {
      id: 'bundle04',
      name: 'Unfinished Bundle',
      task: 'UPLOAD',
      status: 'DRAFT'
    },
    {
      id: 'bundle05',
      name: 'Unfinished Video Bundle',
      task: 'UPLOAD',
      status: 'DRAFT'
    },
    {
      id: 'bundle06',
      name: 'DBL Bundle',
      task: 'DOWNLOAD',
      status: 'NOT_STARTED'
    },
    {
      id: 'bundle07',
      name: 'DBL Bundle 3',
      task: 'DOWNLOAD',
      status: 'NOT_STARTED'
    },
    {
      id: 'bundle08',
      name: 'Audio Bundle #2',
      revision: 40,
      task: 'DOWNLOAD',
      status: 'COMPLETED'
    },
    {
      id: 'bundle09',
      name: 'Audio Bundle #3',
      revision: 5,
      task: 'SAVETO',
      status: 'IN_PROGRESS',
      progress: 0,
    },
    {
      id: 'bundle10',
      name: 'Audio Bundle #4',
      revision: 4,
      task: 'SAVETO',
      status: 'IN_PROGRESS',
      progress: 66,
    },
    {
      id: 'bundle11',
      name: 'Audio Bundle #5',
      revision: 5,
      task: 'SAVETO',
      status: 'COMPLETED',
      progress: 100,
    }
  ];
  const taskOrder = ['UPLOAD', 'DOWNLOAD', 'SAVETO'];
  const statusOrder = ['IN_PROGRESS', 'DRAFT', 'COMPLETED', 'NOT_STARTED'];
  const sortedBundles = sort(bundles).asc([
    b => statusOrder.indexOf(b.status),
    b => taskOrder.indexOf(b.task),
    b => b.name
  ]);
  return sortedBundles;
}
