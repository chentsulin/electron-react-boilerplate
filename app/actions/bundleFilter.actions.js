import { bundleFilterConstants } from '../constants/bundleFilter.constants';

export const bundleFilterActions = {
  updateSearchInput,
  addSearchMatch,
  clearSearch
};

export default bundleFilterActions;

export function updateSearchInput(searchInput, bundles) {
  if (searchInput.length > 0) {
    return { type: bundleFilterConstants.UPDATE_SEARCH_INPUT, searchInput, bundles };
  }
  return clearSearch();
}

function addSearchMatch(bundle) {
  return { type: bundleFilterConstants.ADD_SEARCH_MATCH, bundle };
}

function clearSearch() {
  return { type: bundleFilterConstants.CLEAR_SEARCH_RESULTS };
}
