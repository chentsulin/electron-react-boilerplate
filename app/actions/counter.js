export const INCREMENT_ASYNC = 'INCREMENT_ASYNC';
export const INCREMENT_IF_ODD = 'INCREMENT_IF_ODD';
export const INCREMENT_COUNTER = 'INCREMENT_COUNTER';
export const DECREMENT_COUNTER = 'DECREMENT_COUNTER';

export function increment() {
  return {
    type: INCREMENT_COUNTER
  };
}

export function decrement() {
  return {
    type: DECREMENT_COUNTER
  };
}

export function incrementIfOdd() {
  return {
    type: INCREMENT_IF_ODD
  };
}

export function incrementAsync() {
  return {
    type: INCREMENT_ASYNC
  };
}
