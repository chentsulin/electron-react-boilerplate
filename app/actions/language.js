/*
 *
 * LanguageProvider actions
 *
 */
export const CHANGE_LOCALE = 'CHANGE_LOCALE';

export function changeLocale(languageLocale) {
  return {
    type: CHANGE_LOCALE,
    locale: languageLocale
  };
}
