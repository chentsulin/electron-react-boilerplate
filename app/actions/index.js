import * as globalActions from './global';
import * as galleryActions from './gallery';
import * as languageActions from './language';

const actions = {
  ...globalActions,
  ...galleryActions,
  ...languageActions,
};

export default actions;
