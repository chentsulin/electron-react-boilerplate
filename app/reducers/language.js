import { fromJS } from 'immutable';

import { CHANGE_LOCALE } from '../actions/language';
import { DEFAULT_LOCALE } from '../i18n';

export const initialState = fromJS({
  locale: DEFAULT_LOCALE
});

export default function language(state = initialState, action) {
  switch (action.type) {
    case CHANGE_LOCALE:
      return state.set('locale', action.locale);
    default:
      return state;
  }
}
