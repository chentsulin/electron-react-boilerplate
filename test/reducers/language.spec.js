import Imutable from 'immutable';
import languageReducer from '../../app/reducers/language';
import { CHANGE_LOCALE } from '../../app/actions/language';

describe('reducers', () => {
  describe('gallery', () => {
    it('should handle initial state', () => {
      expect(languageReducer(undefined, {})).toMatchSnapshot();
    });

    it('should handle CHANGE_LOCALE', () => {
      const state = Imutable.Map({});

      const action = {
        type: CHANGE_LOCALE,
        locale: 'en',
      };

      expect(languageReducer(state, action)).toMatchSnapshot();
    });

    it('should handle unknown action type', () => {
      const state = Imutable.Map({});

      expect(languageReducer(state, { type: 'unknown' })).toMatchSnapshot();
    });
  });
});
