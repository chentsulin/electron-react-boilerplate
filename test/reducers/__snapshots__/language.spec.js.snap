// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`reducers gallery should handle CHANGE_LOCALE 1`] = `
Immutable.Map {
  "locale": "en",
}
`;

exports[`reducers gallery should handle initial state 1`] = `
Immutable.Map {
  "locale": "en",
}
`;

exports[`reducers gallery should handle unknown action type 1`] = `Immutable.Map {}`;
